src/libguac/protocol.c

```diff
@@ -74,6 +75,141 @@ ssize_t __guac_socket_write_length_double(guac_socket* socket, double d) {
 
 }
 
+/**
+ * Converts the image data on a Cairo surface to a JPEG image and sends it
+ * as a base64-encoded transmission on the socket.
+ * 
+ * @param socket
+ *     The guac_socket connection to use.
+ * 
+ * @param surface
+ *     A cairo surface containing the image data to send. 
+ * 
+ * @param quality
+ *     JPEG image quality.
+ * 
+ * @return
+ *     Zero on success, non-zero on error.
+ */
+static int __guac_socket_write_length_jpeg(guac_socket* socket, 
+        cairo_surface_t* surface, int quality) {
+
+    /* Get image surface properties and data */
+    cairo_format_t format = cairo_image_surface_get_format(surface);
+
+    if (format != CAIRO_FORMAT_RGB24) {
+        guac_error = GUAC_STATUS_INTERNAL_ERROR;
+        guac_error_message = "Invalid Cairo image format. Unable to create JPEG.";
+        return -1;
+    }
+
+    int width = cairo_image_surface_get_width(surface);
+    int height = cairo_image_surface_get_height(surface);
+    int stride = cairo_image_surface_get_stride(surface);
+    unsigned char* data = cairo_image_surface_get_data(surface);
+
+    /* Flush pending operations to surface */
+    cairo_surface_flush(surface);
+
+    /* Prepare JPEG bits */
+    struct jpeg_compress_struct cinfo;
+    struct jpeg_error_mgr jerr;
+    cinfo.err = jpeg_std_error(&jerr);
+    jpeg_create_compress(&cinfo);
+
+    unsigned char *mem = NULL;
+    unsigned long mem_size = 0;
```

Comment:
Yes, it should be set properly so you can remove the initialization.
I usually like to initialize variables regardless.

On Wed, Aug 5, 2015 at 3:26 PM, Mike Jumper <notifications@github.com>
wrote:

> In src/libguac/protocol.c
> <https://github.com/glyptodon/guacamole-server/pull/63#discussion_r36362282>
> :
>
> > +    int width = cairo_image_surface_get_width(surface);
> > +    int height = cairo_image_surface_get_height(surface);
> > +    int stride = cairo_image_surface_get_stride(surface);
> > +    unsigned char* data = cairo_image_surface_get_data(surface);
> > +
> > +    /* Flush pending operations to surface */
> > +    cairo_surface_flush(surface);
> > +
> > +    /* Prepare JPEG bits */
> > +    struct jpeg_compress_struct cinfo;
> > +    struct jpeg_error_mgr jerr;
> > +    cinfo.err = jpeg_std_error(&jerr);
> > +    jpeg_create_compress(&cinfo);
> > +
> > +    unsigned char *mem = NULL;
> > +    unsigned long mem_size = 0;
>
> @flangelo <https://github.com/flangelo> - this looks like unnecessary
> initialization, but I'm not familiar enough with libjpeg to say for sure.
> Am I safe in assuming that it's OK to leave mem and mem_size unset here,
> and rely on jpeg_mem_dest() to set them properly?
>
> If so, I'll make that change real quick.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/glyptodon/guacamole-server/pull/63/files#r36362282>.
>